src/main/java/org/squiddev/cobalt/function/LuaInterpreter.java

```diff
@@ -640,6 +655,822 @@ A C R(A+3), ... ,R(A+2+C):= R(A)(R(A+1),
 		}
 	}
 
+	private static Function<UnwindableRunnable, UnwindableCallable> continuation(LuaState state, int pc, Prototype proto) {
+		return f -> {
+			final UnwindableCallable callable = (di, c) -> {
+				f.run(di);
+				c.programCounter = di.pc + 1;
+			};
+
+			proto.compiledInstrs[pc] = callable;
+			return callable;
+		};
+	}
+
+	private static Function<UnwindableCallable, UnwindableCallable> rawCont(LuaState state, int pc, Prototype proto) {
+		return raw -> {
+			final UnwindableCallable f = (di, c) -> {
+				raw.call(di, c);
+			};
+
+			proto.compiledInstrs[pc] = f;
+			return f;
+		};
+	}
+
+	static Varargs partialEval(final LuaState state, DebugFrame di, LuaInterpretedFunction function) throws LuaError, UnwindThrowable {
+		Prototype proto = function.getPrototype();
+		EvalCont cont = new EvalCont();
+		final DebugHandler handler = state.debug;
+		DebugState ds = DebugHandler.getDebugState(state);
+
+		while (true) {
+			handler.onInstruction(ds, di, di.pc);
+			final UnwindableCallable compiledInstr = proto.compiledInstrs[di.pc];
+			if (compiledInstr != null) {
+				compiledInstr.call(di, cont);
+			} else {
+				partialEvalStep(state, di, function).call(di, cont);
+			}
+
+			if (cont.varargs != null) return cont.varargs;
+			if (cont.debugFrame != null) {
+				di = cont.debugFrame;
+				cont.programCounter = di.pc;
+				function = cont.function;
+				proto = function.getPrototype();
+				ds = DebugHandler.getDebugState(state);
+				cont.debugFrame = null;
+				cont.function = null;
+			}
+
+			di.pc = cont.programCounter;
+		}
+	}
+
+	// FIXME beware: if a reference to initialFrame makes it into any of the lambdas it could introduce serious memory leaks
```

Comment:
Can this be fixed now? I assume it'd be quite easy to just pass the pc and `Prototype`? There's also a bit of me which wonders if we want to shift this (and any partial evaluation stuff) to an external class?